cpp/src/umap/simpl_set_embed/algo.cuh

```diff
@@ -103,6 +107,50 @@ inline void optimization_iteration_finalization(UMAPParams *params,
   seed += 1;
 }
 
+/**
+ * \brief Constructs a rounding factor used to truncate elements in a sum such that the
+ * sum of the truncated elements is the same no matter what the order of the sum is.
+ *
+ * Algorithm 5: Reproducible Sequential Sum in 'Fast Reproducible Floating-Point
+ * Summation' by Demmel and Nguyen
+ *
+ * In algorithm 5 the bound is calculated as $max(|v_i|) * n$.  We use maximum number of
+ * edges connected to each vertex as n.
+ *
+ * The calculation trick is borrowed from fbcuda, which is BSD-licensed.
+ */
+template <typename T>
+T create_rounding_factor(T max_abs, int n) {
+  T delta =
+    max_abs / (static_cast<T>(1.0) - 2 * n * std::numeric_limits<T>::epsilon());
+
+  // Calculate ceil(log_2(delta)).
+  // frexpf() calculates exp and returns `x` such that
+  // delta = x * 2^exp, where `x` in (-1.0, -0.5] U [0.5, 1).
+  // Because |x| < 1, exp is exactly ceil(log_2(delta)).
+  int exp;
+  std::frexp(delta, &exp);
+
+  // return M = 2 ^ ceil(log_2(delta))
+  return std::ldexp(static_cast<T>(1.0), exp);
+}
+
+template <typename T>
+float create_gradient_rounding_factor(
+  const int *head, int nnz, int n_samples, T alpha,
+  std::shared_ptr<raft::mr::device::allocator> d_alloc, cudaStream_t &stream) {
+  MLCommon::device_buffer<T> buffer(d_alloc, stream, n_samples);
```

Comment:
Can we do it in a different PR?  It seems it's best to use one type of allocation consistently.

Reply:
I'd be inclined to agree if this change weren't so trivial to implement. I'd prefer to start using it since it's being used underneath the raft device buffer as well and the API is almost exactly the same.

Here's a tiny example:
```c++
#include <rmm/device_uvector.hpp>
rmm::device_uvector<T> buffer(n_samples, stream);
```

The allocator is implicit in rmm (again, same allocator being used by `raft::mr::device::allocator`) so we no longer need to specify that.

Reply:
Thanks for the example.  I converted 3 new allocations into `uvector`.